banman: save the banlist in a JSON format on disk

vasild opened the pull request:
Save the banlist in `banlist.json` instead of `banlist.dat`.

This makes it possible to store Tor v3 entries in the banlist on disk
(and any other addresses that cannot be serialized in addrv1 format).

Only read `banlist.dat` if it exists and `banlist.json` does not exist (first start after an upgrade).

Supersedes https://github.com/bitcoin/bitcoin/pull/20904
Resolves https://github.com/bitcoin/bitcoin/issues/19748

vasild committed:
test: ensure banlist can be read from disk after restart

With `banlist.dat` (being written in addrv1 format) if we would try to
write a Tor v3 subnet, it would serialize as a dummy-all-0s IPv6
address and subsequently, when deserialized will not result in the same
subnet.

This problem does not exist with `banlist.json` where the data is saved
in textual, human-readable form.

vasild committed:
banman: save the banlist in a JSON format on disk

Save the banlist in `banlist.json` instead of `banlist.dat`.

This makes it possible to store Tor v3 entries in the banlist on disk
(and any other addresses that cannot be serialized in addrv1 format).

Only read `banlist.dat` if it exists and `banlist.json` does not
exist (first start after an upgrade).

Supersedes https://github.com/bitcoin/bitcoin/pull/20904
Resolves https://github.com/bitcoin/bitcoin/issues/19748

Dusto-beep reviewed (changes requested):
(off topic reply deleted)

theStack commented:
Concept ACK

vasild commented:
790727eea..fd751850e: add the tests from #20904, they are precious

vasild commented:
fd751850e..2d287f639: add release notes

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -105,6 +105,15 @@ RPC
+  in JSON format in `banlist.json` instead of `banlist.dat`. Both files are
+  read on startup and contents merged. Updateds are only written to the new
+  `banlist.json`. A future version of Bitcoin Core may completely ignore
+  `banlist.dat`. (#20966)
```
Please add the release notes to https://github.com/bitcoin-core/bitcoin-devwiki/wiki/%2320852-Release-notes-snippet

Otherwise they will have to be removed here after merge, then added to the `0.21` branch. (Release notes are only added to the point releases, not the next major version that also includes those fixes)

vasild replied:
I guess after/if this PR is merged?

laanwj commented:
Concept ACK

vasild commented:
2d287f639...fd751850e: remove edits of `doc/release-notes.md`

vasild commented:
Notes:

* I guess it would be possible to write the ban list to `settings.json` instead of in `banlist.json` (as originally proposed in https://github.com/bitcoin/bitcoin/issues/19748), but the periodic dump may create some complications and also it is unclear what to do if we are running with `-nosettings`.

* The current patch would write `banlist.json` on the first change of the ban list. This means that a node that upgrades from before this PR with only `banlist.dat` on disk may not create `banlist.json` for some time. I think this is ok.

promag reviewed (commented):
Concept ACK.

Should try to read JSON first, and if it exists then skip old format? Or am I missing something?

Personally I wouldn't  add from/to JSON in `CBanEntry`, `BanMapFromJson` and `BanMapToJson` seem enough. Also, there should be some validation in `BanMapFromJson`?

jonatack commented:
Concept ACK

> [790727e](https://github.com/bitcoin/bitcoin/commit/790727eeaf165449f691d1f874b70f0816ef7601)..[fd75185](https://github.com/bitcoin/bitcoin/commit/fd751850e5dc918b523c63ce9db98279f2398c96): add the tests from #20904 (closed without merge), they are precious

Good to see these tests brought in.

vasild commented:
fd751850e...e74196741: read `banlist.dat` only if `banlist.json` is not present and catch all exceptions that could come from an invalid JSON.

@promag  
> Should try to read JSON first, and if it exists then skip old format?

I made it to read both if present so that if the user temporary downgrades and makes some changes to `banlist.dat` those changes would be picked up by the new version. However this has the deficiency that if an entry is deleted from `banlist.json` it will keep being loaded from `banlist.dat`. Changed to what you suggest above.

> Personally I wouldn't add from/to JSON in `CBanEntry`, `BanMapFromJson` and `BanMapToJson` seem enough.

`CBanEntry` being able to ser/unser itself to/from JSON improves its encapsulation.

> Also, there should be some validation in `BanMapFromJson`?

Right! Added `try/catch` to handle invalid JSONs.

MarcoFalke commented:
needs rebase after #21167

vasild commented:
`e74196741...7493bbe38`: rebase due to conflicts

trongham commented on src/test/util/setup_common.cpp:
```diff
@@ -187,7 +187,7 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
         throw std::runtime_error(strprintf("ActivateBestChain failed. (%s)", state.ToString()));
     }
 
-    m_node.banman = MakeUnique<BanMan>(GetDataDir() / "banlist.dat", nullptr, DEFAULT_MISBEHAVING_BANTIME);
```
-    m_node.banman = MakeUnique<BanMan>(GetDataDir() / "banlist", nullptr, DEFAULT_MISBEHAVING_BANTIME) ;

MarcoFalke commented:
So many rebases. I hope the backport won't be too nasty

vasild commented:
`1310b5460...469da8ca3`: rebase due to conflicts

jonatack commented:
Concept ACK

MarcoFalke commented on src/addrdb.cpp:
```diff
@@ -119,18 +175,46 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
+    }
+
+    for (const auto& err : errors) {
+        error("%s: %s", __func__, err);
```
same commit:

* It seems the caller ignores the return code, so calling this ERROR seems confusing
* No longer need `__func__`

vasild replied:
The caller of `CBanDB::Write()` does not ignore its return code? If you mean the return code of `error()` itself - this is because we want to log all of the errors (possibly more than one) before returning `false`.

Removed `__func__`.

MarcoFalke commented on src/addrdb.cpp:
```diff
@@ -119,18 +175,46 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
+
+    if (!util::ReadSettings(m_banlist_json, settings, errors)) {
+        for (const auto& err : errors) {
+            error("%s: %s", __func__, err);
```
same commit:

* The caller will be "Recreating the banlist database", so calling this ERROR seems confusing
* `__func__` ;)

vasild replied:
I think it is ok to see something like the following in the log:
```
ERROR: ... balist foo ...
ERROR: ... banlist bar ...
Recreating the banlist database
```

Removed `__func__`.

MarcoFalke replied:
User will report bugs if they see `ERROR`. For example, http://www.erisian.com.au/bitcoin-core-dev/log-2021-02-13.html#l-296

vasild replied:
I changed it to avoid the word "error". I don't think those errors should be hushed - those are errors from loading the file and we already checked that the file exists. So may include something like "json parse error", "io error", "permission denied".

In general I don't think we should avoid using "error" or use it just for cases where we have detected a bug in the program. Messages like "error: permission denied" or "error: connection reset" are to be expected.

vasild replied:
I am leaving it as is, I don't like `error()` too much.

MarcoFalke commented on src/addrdb.cpp:
```diff
@@ -119,18 +175,46 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
+    try {
+        return BanMapFromJson(settings[JSON_KEY], banSet);
+    } catch (const std::runtime_error& e) {
+        return error("%s: Cannot parse %s: %s", __func__, m_banlist_json.string(), e.what());
```
same ;)

vasild replied:
Removed `__func__`.

MarcoFalke commented on src/test/fuzz/banman.cpp:
```diff
@@ -34,8 +34,9 @@ FUZZ_TARGET_INIT(banman, initialize_banman)
-    const fs::path banlist_file = GetDataDir() / "fuzzed_banlist.dat";
-    fs::remove(banlist_file);
+    const fs::path banlist_file = GetDataDir() / "fuzzed_banlist";
+    fs::remove(banlist_file.string() + ".dat");
```
same commit:

I don't understand this line. There shouldn't be any reason this instance will ever write a banlist.dat file.

Also, to make the fuzz test more useful, it should probably *create* a banlist.dat file. Extra points for creating a corrupted one. With extra bonus points for passing one in without disk access.

vasild replied:
Right, removed the `fs::remove` calls and replaced them with a creation of a corrupted or inaccessible banlist file.

maflcko reviewed (approved):
review ACK 469da8ca3f566041972a87e847e7ecf44ef2306b 🎐

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 469da8ca3f566041972a87e847e7ecf44ef2306b 🎐
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiXDwv9E2HLRr4rmJmOP/mO+47lhG6gEcqx6EAYuIjlBklEnDK9PhDo43/t3AzN
suyNHfsJ9mGGD/QXoZxUYge2dfqaH1U1y/BdrlVZw02r/HXqvOGzJ4YMjV4ysw04
vVh0U1osD7w3ypYLR0BhcAj4SVR1NDT3SYMKf8qlfPFtXfrn6W9i0kG6siwdV6ZE
mw/IqbisqDq92azMw+26z07Mu7uXv4Z1zWiz0qD7KBsJ/pinDdXCNi45gde7G091
9tMF7KOJYyv/6uY/ibXoq/MYSDvkwye0SgGkNh6kku1MBjscBZG2KF/B6YIZMOn7
2GTl+vRyUWMmsmANB7ELHyJs0wlh30bNGpjUErWfYwH4fs7KKLiQfvohms3UFOI4
oLeAsqpDo8HX8X0SzMmIOib/bhg6+X5yxIJW113pLfuohc35bBKEHATvZlzmYjSZ
lXFVLKY38YKsbxCBoXr3ISL7OeKKn6AwFVFZLjA8BlbogASWI9bunmJyHuC/o1n2
n2ha9e1V
=sn/g
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `4f0238a21835acd2f6c1095be832c4be7733461a6b3f0bf1498f5bab2c98df7d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401084f0238a21835acd2f6c1095be832c4be7733461a6b3f0bf1498f5bab2c98df7df0105d70040f52450ff8ae83582e546e90ce08fff0103919403cb6f75779690036e620116e8c08f1046055c01ff008e6e6d19a1961b9c70083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01056621442f78798f65e433df07c0b9a8108f1046055c01ff008f7458e0bdb2936a40083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0109370c6a740be139db9cc74affd33793708f1046055b249f008f9644eeeeef47deb0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010bf8dab80e457ff3b3355103e959e6c0e08f1046055c01ff0081044a299be64cd3f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

vasild commented:
`469da8ca3...4efa5e73b`: address review suggestions, remove the release notes that were removed before but sneaked back in one of the updates

vasild commented on src/addrdb.cpp:
```diff
@@ -11,13 +11,69 @@
+UniValue BanMapToJson(const banmap_t& bans)
+{
+    UniValue bans_json(UniValue::VARR);
+    for (const auto& [address, ban_entry] : bans) {
```
My understanding is that this deserves a backport because it would allow storing torv3 addresses in the banlist on disk. Not super important though.

Switched to C++11.

vasild commented on src/addrdb.cpp:
```diff
@@ -11,13 +11,69 @@
+        CSubNet subnet;
+        const auto& subnet_str = ban_entry_json["address"].get_str();
+        if (!LookupSubNet(subnet_str, subnet)) {
+            return error("%s: Cannot parse banned address or subnet: %s", __func__, subnet_str);
```
You probably mean `-logsourcelocations` instead of `-logfunctionnames`.

Removed `__func__`. Changed this function to throw an exception for all errors and return `void`. Now it does not do any logging - just bubble up a detailed error message to the callers via the exception.

MarcoFalke commented:
@vasild I've split up the unit tests from this pull, since they are unrelated to changes here (They are checking functionality that already works just fine on master). See #21571. The benefit is that this pull can focus on the disk serialization changes only. I hope you don't mind and maybe can leave a (Concept) (N)ACK.

MarcoFalke commented:
Rebase should be easy now `git rebase --skip`

vasild commented:
`4efa5e7...03239a3ab`: rebase due to conflicts, remove 3 commits from this PR that were already merged via https://github.com/bitcoin/bitcoin/pull/21571.

laanwj commented:
Moving milestone to 0.21.2. We'd like to do 0.21.1 soon, and even though this may be ready for merge soon, it seems like a risky thing to rush into a point release.

MarcoFalke commented on src/banman.cpp:
```diff
@@ -188,7 +188,7 @@ void BanMan::SweepBanned()
                 m_is_dirty = true;
                 notify_ui = true;
-                LogPrint(BCLog::NET, "%s: Removed banned node ip/subnet from banlist.dat: %s\n", __func__, sub_net.ToString());
+                LogPrint(BCLog::NET, "%s: Removed banned node address/subnet: %s\n", __func__, sub_net.ToString());
```
Can remove `__func__` while touching this line?

vasild commented:
`03239a3ab...aa90a7d46`: address suggestions

promag commented on src/addrdb.cpp:
```diff
@@ -119,18 +176,49 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
 bool CBanDB::Read(banmap_t& banSet)
 {
-    return DeserializeFileDB(m_ban_list_path, banSet);
+    if (!fs::exists(m_banlist_json)) {
```
51909ded1e8b26158f8d36cfe323ac72feac1d78

A comment here would be nice. Or maybe just take the next suggestion so it's clear what this is doing.

vasild replied:
Added some comments.

promag commented on src/addrdb.cpp:
```diff
@@ -119,18 +176,49 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
 {
-    return DeserializeFileDB(m_ban_list_path, banSet);
+    if (!fs::exists(m_banlist_json)) {
+        return DeserializeFileDB(m_banlist_dat, banSet);
```
51909ded1e8b26158f8d36cfe323ac72feac1d78

Maybe migrate on 1st run?
```diff
 bool CBanDB::Read(banmap_t& banSet)
 {
     if (!fs::exists(m_banlist_json)) {
-        return DeserializeFileDB(m_banlist_dat, banSet);
+        if (!DeserializeFileDB(m_banlist_dat, banSet)) {
+            return false;
+        }
+        // log something
+        Write(banSet);
+        return true;
     }
```

vasild replied:
Migrate on first run - yes. But `Read()` calling `Write()` seems like some anti-pattern to me. Instead I signal from `Read()` that the list needs re-flushing, so that the caller takes care of it.

promag replied:
Good point. Maybe add a `CBanDB::MigrateAndRead` where calling `Write()` isn't so bad.

vasild replied:
Not necessary with the latest changes.

promag reviewed (commented):
Tested ACK 03239a3ab204dfc71bb46f85679414b331836763. Left some comments for your consideration.

vasild commented:
`aa90a7d46...7e399e2de`: rebase due to conflicts

vasild commented:
`7e399e2de...08f4f98ea`: address some suggestions, force saving on disk if we loaded just the old `banlist.dat` so that we create the new `banlist.json` this way, without waiting for mods to the list in order to flush it to disk.

vasild commented on test/functional/rpc_setban.py:
```diff
@@ -51,12 +51,22 @@ def run_test(self):
         assert(self.is_banned(node, tor_addr))
         assert(not self.is_banned(node, ip_addr))
+
+        self.restart_node(1)
```
Added a comment.

vasild commented on src/addrdb.cpp:
```diff
@@ -119,18 +176,49 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
 
-CBanDB::CBanDB(fs::path ban_list_path) : m_ban_list_path(std::move(ban_list_path))
+CBanDB::CBanDB(fs::path ban_list_path)
+    : m_banlist_dat(ban_list_path.string() + ".dat"),
```
I find `replace_extension()` a bit confusing in this case because we know that there is no extension, so we are not replacing anything (I know it will do the right thing, but still).

jonatack reviewed (commented):
Reviewing and testing this later today.

jonatack commented:
ACK 08f4f98ea2cdc3fbff5c9e21870296106430e468 with the suggestion to perhaps remove any leftover "::/0" addresses from mis-saved addrv2 ban entries when creating the json file and add a test. They will eventually expire, so this could be done in a follow-up, but it may be good to avoid this confusing case as banlist.json is trivially user-readable. Tested on master that an existing banlist.dat containing 3 IPv4, 2 torv3, 1 I2P is read on restart (with an entry of "::/0" or an "IPv4" address for the addv2 addresses), and that after building this patch, a banlist.json file with the existing entries is auto-created on startup except with incorrect ban entries for the addrv2 addresses, and that subsequent setban add/remove operations correctly update the `banlist.json` file for torv3 and i2p addresses. As expected, attempting to remove any previously incorrectly saved addrv2 addresses fails to recognize it unless the "new" address is removed, but subsequent add/remove operations work correctly. This testing was done by simultaneously looking at the banlist files and running setban/listbanned, shutting down, rm-ing the banlist.json file, restarting on master and this patch, and repeating the testing.

vasild committed:
doc: fix grammar in doc/files.md

vasild committed:
style: remove () from assert in rpc_setban.py

vasild commented on doc/files.md:
```diff
@@ -55,7 +55,8 @@ Subdirectory       | File(s)               | Description
 `wallets/`         |                       | [Contains wallets](#multi-wallet-environment); can be specified by `-walletdir` option; if `wallets/` subdirectory does not exist, wallets reside in the [data directory](#data-directory-location)
 `./`               | `anchors.dat`         | Anchor IP address database, created on shutdown and deleted at startup. Anchors are last known outgoing block-relay-only peers that are tried to re-connect to on startup
-`./`               | `banlist.dat`         | Stores the IPs/subnets of banned nodes
+`./`               | `banlist.dat`         | Stores the addresses/subnets of banned nodes (deprecated). Bitcoin Core will not save the banlist to it. It would only read it on startup if `banlist.json` is not present.
```
Done. Put that last change in a separate commit.

vasild commented on test/functional/rpc_setban.py:
```diff
@@ -51,12 +51,24 @@ def run_test(self):
 
+        self.restart_node(1)
+        assert(not self.is_banned(self.nodes[1], tor_addr))
+        assert(not self.is_banned(node, ip_addr))
```
Done. Since the surrounding code uses `assert()` I added the new code with `assert()` for consistency and then removed all `()` from asserts in the entire file in a separate commit.

vasild commented:
> ... remove any leftover (e.g. "::/0") addresses from mis-saved addrv2 ban entries when creating the json file ...

Hmm. Could there be real, manually added, no-mistake, ban entries like`::/0`? Somebody has banned deliberately the entire IPv6 network?

jonatack commented on test/functional/rpc_setban.py:
```diff
@@ -22,7 +22,7 @@ def run_test(self):
         self.connect_nodes(0, 1)
         peerinfo = self.nodes[1].getpeerinfo()[0]
-        assert(not 'noban' in peerinfo['permissions'])
+        assert not "noban" in peerinfo["permissions"]
```
fe4aafda512483e863b23cfad403509b1604555c drive-by nit, while touching lines 25 and 46 in this commit

`$ pycodestyle test/functional/rpc_setban.py`

    test/functional/rpc_setban.py:25:16: E713 test for membership should be 'not in'
    test/functional/rpc_setban.py:46:16: E713 test for membership should be 'not in'

vasild replied:
Going to leave it as is because none of "black" or "pycodestyle" is mentioned in `doc/`. Don't want to bother other reviewers with too big and unrelated changes to this PR.

jonatack replied:
Ok. For info, pycodestyle is PEP8 that we follow per doc/test/functional/README.md

https://pypi.org/project/pycodestyle

jonatack replied:
(ah, it was renamed from PEP8 for this reason https://github.com/PyCQA/pycodestyle/issues/466)

vasild replied:
`lint-python.sh` is happy with the current `rpc_setban.py`:

```sh
$ ./test/lint/lint-python.sh
test/functional/test_runner.py:42: error: Module has no attribute "getwindowsversion"
Found 1 error in 1 file (checked 199 source files)
$
```

jonatack replied:
If adding new code or retouching old code for style reasons, I'd try to make it conform (https://www.flake8rules.com/rules/E713.html for "not in") because reviewers are often picky about it. I also wouldn't add a commit that does a one-word grammar change because otherwise someone will invariably ask me to squash the commit. But everyone has a different experience with this so definitely feel free to ignore.

jonatack commented:
> > ... remove any leftover (e.g. "::/0") addresses from mis-saved addrv2 ban entries when creating the json file ...
> 
> Hmm. Could there be real, manually added, no-mistake, ban entries like `::/0`? Somebody has banned deliberately the entire IPv6 network?

Good question. Would anyone do that. OTOH, have people been banning tor v3 addresses yet. Possible. IDK if it's an edge case worth worrying about.

dhruv reviewed (commented):
ACK `fe4aafd`

Code reviewed and tested locally. Thank you for the patch @vasild ! I especially like that node operators will have an option to share banlists as a human-readable json file instead of rpc commands 🚀

To verify things work as intended:
```shell
$ git pull --rebase master
$ make -j 9
$ mkdir -p node0 node1

# Run node 0
$ src/bitcoind -regtest -datadir=`pwd`/node0 -rpccookiefile=.cookie -port=2222 -rpcport=3333

# Run node 1
$ src/bitcoind -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -port=2223 -rpcport=3334

# Node 0 adds Node 1 as peer
$ src/bitcoin-cli -regtest -datadir=`pwd`/node0 -rpccookiefile=.cookie -rpcport=3333 addnode "127.0.0.1:2223" "onetry"

# Node 1 sees the peer
$ src/bitcoin-cli -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -rpcport=3334 getpeerinfo

#Node 1 bans Node 0
$ src/bitcoin-cli -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -rpcport=3334 setban "127.0.0.1" "add"
$ src/bitcoin-cli -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -rpcport=3334 listbanned

# Node 1 eliminates the peer
$ src/bitcoin-cli -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -rpcport=3334 getpeerinfo

# Stop both nodes manually

# Banlist files on each node
$ ls -lan node0/regtest | grep banlist
-rw-------   1 501  20      37 May 14 11:58 banlist.dat

$ ls -lan node1/regtest | grep banlist
-rw-------   1 501  20      91 May 14 12:08 banlist.dat

$ git checkout -b pr-20966 vasild/json_bans
$ make -j 9

# Run node 0
$ src/bitcoind -regtest -datadir=`pwd`/node0 -rpccookiefile=.cookie -port=2222 -rpcport=3333

# Run node 1
$ src/bitcoind -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -port=2223 -rpcport=3334

# Banlist files on each node
$ ls -lan node0/regtest | grep banlist
-rw-------   1 501  20      37 May 14 11:58 banlist.dat
-rw-------   1 501  20      34 May 14 12:21 banlist.json

$ ls -lan node1/regtest | grep banlist
-rw-------   1 501  20      91 May 14 12:08 banlist.dat
-rw-------   1 501  20     167 May 14 12:21 banlist.json

$ cat node0/regtest/banlist.json
{
    "banned_nets": [
    ]
   }

$ cat node1/regtest/banlist.json
{
    "banned_nets": [
     {
      "version": 1,
      "ban_created": 1621019336,
      "banned_until": 1621105736,
      "address": "127.0.0.1/32"
     }
    ]
   }
 
# Unban node 0 from node 1
$ src/bitcoin-cli -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -rpcport=3334 setban "127.0.0.1" "remove"

# Verify banlist.json updated
$ cat node1/regtest/banlist.json
{
    "banned_nets": [
    ]
   }
 
# Manually restart node1
# Verify unchanged banlist.json
$ cat node1/regtest/banlist.json
{
    "banned_nets": [
    ]
   }
```

However, this is the behavior I'd like to point to and make sure it's ok:
```
# Manually stop node1
# Node 1 deletes banlist.json to clear all bans
$ rm node1/regtest/banlist.json
# Restart node 1

# The ban comes back from banlist.dat
$ src/bitcoin-cli -regtest -datadir=`pwd`/node1 -rpccookiefile=.cookie -rpcport=3334 listbanned
[
  {
    "address": "127.0.0.1/32",
    "ban_created": 1621019336,
    "banned_until": 1621105736,
    "ban_duration": 86400,
    "time_remaining": 85186
  }
]

$ cat node1/regtest/banlist.json
{
    "banned_nets": [
     {
      "version": 1,
      "ban_created": 1621019336,
      "banned_until": 1621105736,
      "address": "127.0.0.1/32"
     }
    ]
   }
```

Some other suggestions below.

vasild commented on src/addrdb.cpp:
```diff
@@ -11,13 +11,70 @@
+    for (const auto& ban_entry_json : bans_json.getValues()) {
+        CSubNet subnet;
+        const auto& subnet_str = ban_entry_json["address"].get_str();
+        if (!LookupSubNet(subnet_str, subnet)) {
```
I am not sure. `version` and `banned_until` are not sanity-checked when reading `banlist.dat`, so I do the same with `banlist.json` as the purpose of this PR is to switch the format, not introduce new checks.

Maybe worth considering separately.

vasild commented on src/addrdb.cpp:
```diff
@@ -119,18 +176,54 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
-    return SerializeFileDB("banlist", m_ban_list_path, banSet);
+    std::vector<std::string> errors;
+    if (util::WriteSettings(m_banlist_json, {{JSON_KEY, BanMapToJson(banSet)}}, errors)) {
+        return true;
```
I deliberately avoid removing `banlist.dat` automatically because that would prevent downgrading to an older version (and keeping the ban list).

Something like this may be a good idea:

```cpp
--- i/src/addrdb.cpp
+++ w/src/addrdb.cpp
@@ -185,12 +185,18 @@ CBanDB::CBanDB(fs::path ban_list_path)
 }

 bool CBanDB::Write(const banmap_t& banSet)
 {
     std::vector<std::string> errors;
     if (util::WriteSettings(m_banlist_json, {{JSON_KEY, BanMapToJson(banSet)}}, errors)) {
+        if (fs::exists(m_banlist_dat)) {
+            LogPrintf("Successfully saved %s to disk. %s can be removed manually now if "
+                      "downgrading to an older version is not planned.\n",
+                      m_banlist_json.string(),
+                      m_banlist_dat.string());
+        }
         return true;
     }

     for (const auto& err : errors) {
         error("%s", err);
     }
```

but it would be printed **every** time which would be annoying. There is some doc added to `doc/files.md` which should clarify things if the user wonders why there are two files. Maybe leave it as is?

We can add an automatic removal of `banlist.dat` one major version after this change is released.

dhruv replied:
Removing `banlist.dat` one major version later seems like a fair choice!

vasild commented on src/banman.h:
```diff
@@ -17,7 +17,8 @@
 static constexpr unsigned int DEFAULT_MISBEHAVING_BANTIME = 60 * 60 * 24; // Default 24-hour ban
-// How often to dump addresses to banlist.dat
+
+/// How often to dump banned addresses/subnets to disk.
```
The 3 slashes make it a [doxygen comment](https://www.doxygen.nl/manual/docblocks.html#specialblock) for inclusion in https://doxygen.bitcoincore.org/.

jonatack replied:
Yes, and doc/developer-notes.md also provides an example with /// iirc.

dhruv replied:
Oh. Thanks for teaching.

vasild commented on src/test/fuzz/banman.cpp:
```diff
@@ -34,8 +36,20 @@ FUZZ_TARGET_INIT(banman, initialize_banman)
     SetMockTime(ConsumeTime(fuzzed_data_provider));
-    const fs::path banlist_file = GetDataDir() / "fuzzed_banlist.dat";
-    fs::remove(banlist_file);
+    fs::path banlist_file = GetDataDir() / "fuzzed_banlist";
```
I don't see how - we must first write the file or set `banlist_file` and only after that execute the actual test.

dhruv replied:
Perhaps with a separate fuzz test? I'm a fuzzing n00b though. We should ask @MarcoFalke and @practicalswift!

vasild commented on src/addrdb.h:
```diff
@@ -62,15 +76,30 @@ class CAddrDB
+     * Read the banlist from disk.
+     * @param[out] banSet The loaded list, set if `true` is returned, otherwise it is left
+     * in an undefined state.
+     * @param[out] dirty Indicates whether the loaded list needs flushing to disk, only set
```
Right, reworded the comment.

vasild commented on src/addrdb.cpp:
```diff
@@ -11,13 +11,70 @@
+        const auto& address = it.first;
+        const auto& ban_entry = it.second;
+        UniValue j = ban_entry.ToJson();
+        j.pushKV("address", address.ToString());
```
Yes, good idea! Done.

jonatack replied:
Not a fan of doing this for a short string like this that is used only twice, but it's not a blocker.

vasild commented on src/addrdb.cpp:
```diff
@@ -119,18 +176,54 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
 {
-    return SerializeFileDB("banlist", m_ban_list_path, banSet);
+    std::vector<std::string> errors;
+    if (util::WriteSettings(m_banlist_json, {{JSON_KEY, BanMapToJson(banSet)}}, errors)) {
```
I think either one is fine. Renaming is out of the scope of this PR.

vasild commented:
`fe4aafda51...13c681161c`: address review suggestions

vasild commented:
> this is the behavior I'd like to point to and make sure it's ok ... Node 1 deletes banlist.json to clear all bans ... The ban comes back from banlist.dat

Yes, this is how it is supposed to work. The operator should have seen that there are two files `banlist.dat` and `banlist.json` and deleted both or consulted `doc/files.md`. We don't automatically delete `banlist.dat` after creating `banlist.json` because that would create issues if the user decides to downgrade to the previous version which only recognizes `banlist.dat`.

dhruv commented:
re-ACK `13c6811`

jonatack commented:
Diff-review re-ACK 13c681161c4a3f2a2acd6d812c1360e0ad94019c per ` git diff 08f4f98 13c6811` and  `git range-diff d4c409c 08f4f98 13c6811`

Only non-showstopping mini-regret is the Python "not in" not updated while re-touching to conform to PEP8 per our functional test guidelines.

MarcoFalke commented:
Sorry, this had two ACKs and I didn't see the conflict.

vasild commented:
`13c681161c...88b843e49c`: rebase due to conflicts

<details>
<summary>Release notes</summary>

```diff
--- a/doc/release-notes.md
+++ b/doc/release-notes.md
@@ -182,12 +182,21 @@ RPC
   - `verifymessage` now returns RPC_TYPE_ERROR (-3) if the passed signature
     is malformed. Previously returned RPC_INVALID_ADDRESS_OR_KEY (-5).

 Tests
 -----

+Files
+-----
+
+- The list of banned hosts and networks (via `setban` RPC) is now saved on disk
+  in JSON format in `banlist.json` instead of `banlist.dat`. `banlist.dat` is
+  only read on startup if `banlist.json` is not present. Changes are only
+  written to the new `banlist.json`. A future version of Bitcoin Core may
+  completely ignore `banlist.dat`. (#20966)
+
 Credits
 =======
```
</details>

hebasto commented:
Concept ACK.

jonatack commented on test/functional/rpc_setban.py:
```diff
@@ -36,27 +36,40 @@ def run_test(self):
+
+        self.restart_node(1)
+        assert not self.is_banned(self.nodes[1], tor_addr)
+        assert not self.is_banned(node, ip_addr)
```
f371273 could be done in a follow-up along with `not_in`, but if you retouch
```diff
--- a/test/functional/rpc_setban.py
+++ b/test/functional/rpc_setban.py
@@ -56,18 +56,17 @@ class SetBanTests(BitcoinTestFramework):
         assert self.is_banned(node, tor_addr)
         assert not self.is_banned(node, ip_addr)
 
-        # Test that the ban list is preserved through restart.
-
+        self.log.info("Test the ban list is preserved through restart")
         self.restart_node(1)
         assert self.is_banned(node, tor_addr)
         assert not self.is_banned(node, ip_addr)
 
         node.setban(tor_addr, "remove")
-        assert not self.is_banned(self.nodes[1], tor_addr)
+        assert not self.is_banned(node, tor_addr)
         assert not self.is_banned(node, ip_addr)
 
         self.restart_node(1)
-        assert not self.is_banned(self.nodes[1], tor_addr)
+        assert not self.is_banned(node, tor_addr)
         assert not self.is_banned(node, ip_addr)
```

vasild replied:
Added the log message and changed to `node` in the second one, but left the first one as that line is not modified by this PR.

jonatack replied:
Seems that line was added by me (hides) in 39b43298d9c, so I don't mind if you change it :D

jonatack reviewed (commented):
re-ACK 88b843e49c84e331e40bfbe249269deaf2a6a2ac per `git range-diff ce4a852 13c6811 88b843e`

hebasto commented on doc/files.md:
```diff
@@ -56,7 +56,8 @@ Subdirectory       | File(s)               | Description
 `wallets/`         |                       | [Contains wallets](#multi-wallet-environment); can be specified by `-walletdir` option; if `wallets/` subdirectory does not exist, wallets reside in the [data directory](#data-directory-location)
 `./`               | `anchors.dat`         | Anchor IP address database, created on shutdown and deleted at startup. Anchors are last known outgoing block-relay-only peers that are tried to re-connect to on startup
-`./`               | `banlist.dat`         | Stores the IPs/subnets of banned nodes
+`./`               | `banlist.dat`         | Stores the addresses/subnets of banned nodes (deprecated). `bitcoind` or `bitcoin-qt` no longer save the banlist to this file, but read it on startup if `banlist.json` is not present.
+`./`               | `banlist.json`        | Stores the addresses/subnets of banned nodes.
```
@vasild 
In your opinion, when will be an appropriate moment to move the `banlist.dat` description into the "Legacy subdirectories and files" subsection?

promag replied:
See https://github.com/bitcoin/bitcoin/pull/20966/commits/cd9da1b2a2f2dcad008e5a1a079a093e77565369#r634452821.

vasild replied:
Good question! I am not sure. Either now or in a next release where we completely ignore `banlist.dat`?

promag reviewed (commented):
Code review ACK 88b843e49c84e331e40bfbe249269deaf2a6a2ac.

hebasto commented on test/functional/rpc_setban.py:
```diff
@@ -36,27 +36,40 @@ def run_test(self):
         node.setban(tor_addr, "remove")
-        assert(not self.is_banned(self.nodes[1], tor_addr))
-        assert(not self.is_banned(node, ip_addr))
+        assert not self.is_banned(self.nodes[1], tor_addr)
+        assert not self.is_banned(node, ip_addr)
+
+        self.restart_node(1)
+        assert not self.is_banned(self.nodes[1], tor_addr)
```
As we have https://github.com/bitcoin/bitcoin/blob/88b843e49c84e331e40bfbe249269deaf2a6a2ac/test/functional/rpc_setban.py#L49

what are reasons to use `self.nodes[1]`? Did I miss something?

hebasto replied:
Or, I see https://github.com/bitcoin/bitcoin/pull/20966/files#r637932195

hebasto reviewed (approved):
ACK 88b843e49c84e331e40bfbe249269deaf2a6a2ac, tested on Linux Mint 20.1 including error/exception handling.

vasild commented:
`88b843e49c...4b52c7234f`: rebase due to conflicts

release notes in https://github.com/bitcoin/bitcoin/pull/20966#issuecomment-847001828

jonatack reviewed (commented):
re-ACK 4b52c7234f3c2e3067d7d6c6fd7ebf2b96bb8a45 per `git range-diff 5c4f0c4 88b843e 4b52c72`, rebase only, debug build clean, test passes, repeatedly tested banning/unbanning peers on signet while inspecting banlist.json in my editor including through restarts. However, there are gotchas when downgrading/upgrading; it may be a good idea to warn and suggest best practices about that in the release note.

jonatack commented:
Did a bunch of manual testing and updated my ACK above.

jonatack commented:
> +- The list of banned hosts and networks (via `setban` RPC) is now saved on disk
> +  in JSON format in `banlist.json` instead of `banlist.dat`. `banlist.dat` is
> +  only read on startup if `banlist.json` is not present. Changes are only
> +  written to the new `banlist.json`. A future version of Bitcoin Core may
> +  completely ignore `banlist.dat`. (#20966)

Regarding the release note, peers may be banned and unbanned via the GUI peers window in addition to the setban RPC. As mentioned in my previous comment, it may be a good idea to describe user gotchas and best practices for upgrading and downgrading.

sipa commented:
utACK 4b52c7234f3c2e3067d7d6c6fd7ebf2b96bb8a45

I think it would be useful to make the code ignore future version numbers, so it will not completely wipe banlists on downgrading (from future versions that introduce additional features to this format), but that can be done later.

promag reviewed (commented):
Code review ACK 4b52c7234f3c2e3067d7d6c6fd7ebf2b96bb8a45.

vasild commented:
`4b52c7234f...5fcc96457d`: rebase due to conflicts

vasild commented:
`5fcc96457d...bb719a08db`: minor tweaks in `rpc_setban.py` - log a message and use `node` instead of `self.nodes[1]`

jonatack commented:
Code review re-ACK bb719a08db173a753984a04534de6f148b87b17a per `git range-diff 6a67366 4b52c72 bb719a0`

achow101 commented:
Code Review ACK bb719a08db173a753984a04534de6f148b87b17a

MarcoFalke commented:
Removed from backport, but if someone wants to backport, it will need to be done together with https://github.com/bitcoin/bitcoin/pull/20852#issuecomment-866610187

MarcoFalke commented:
Added to release notes: https://github.com/bitcoin-core/bitcoin-devwiki/wiki/22.0-Release-Notes-draft/_history
